### ISSUE TITLE
Integration of Optuna with Hydra for Hyperparameter Tuning

### ISSUE THREAD

**Issue by dahlem** [open-disease-risk/sat#10]
**Is your feature request related to a problem? Please describe.**
No

**Problem Statement**

  We need a standardized approach to perform hyperparameter tuning using Optuna within our Hydra-based configuration system. Currently,
   there's no clear path for users to leverage Optuna's optimization capabilities while maintaining the benefits of our Hydra
  configuration setup.

**Background**

  - Our project uses Hydra for configuration management
  - Optuna is a powerful hyperparameter optimization framework
  - We need to bridge these two systems for effective parameter tuning

**Proposed Integration**

  Requirements
  1. Allow users to define parameter search spaces in a Hydra-compatible way
  2. Support different optimization objectives (e.g., minimizing loss, maximizing C-index)
  3. Provide integration with our existing logging and experiment tracking
  4. Maintain compatibility with our pipeline structure
  5. Support both local and AzureML execution modes

  Considerations

  - Search Space Definition: How to define Optuna search spaces (categorical, float, int) in Hydra config files
  - Objective Function: How to map model training and evaluation to an Optuna objective
  - Pruning Integration: Support for early stopping of unpromising trials
  - Multi-objective Optimization: For balancing multiple metrics like Brier score and C-index
  - Parallel Execution: Efficient execution of trials, especially on cloud resources

  Proposed Implementation

  We could add an optuna.py module that:
  1. Builds on Hydra's existing sweep capabilities
  2. Defines an Optuna study and trials with configuration from Hydra
  3. Creates a standardized API for defining search spaces
  4. Integrates with MLflow for result tracking

  Example Config (Concept)

```
  # conf/sweep/optuna/metabric.yaml
  defaults:
    - override /experiments: metabric/survival

  optuna:
    study:
      direction: minimize  # or maximize
      metric: test_brier_avg  # which metric to optimize
      n_trials: 50
      sampler: TPESampler
      pruner: MedianPruner

    parameters:
      learning_rate:
        type: float
        low: 1e-5
        high: 1e-3
        log: true

      num_train_epochs:
        type: int
        low: 1
        high: 10

      dropout:
        type: float
        low: 0.1
        high: 0.5

      loss_weight:
        type: categorical
        choices: [0.1, 0.5, 1.0, 2.0, 5.0]
```
  Example Usage

  `python -m sat.optuna experiments=metabric/survival optimizers=optuna`

**Benefits**

  - Standardized approach to hyperparameter tuning
  - Integration with existing metrics and evaluation
  - More efficient model development
  - Better model performance through optimization

**Implementation Questions**

  1. Should we create a new Hydra plugin or use existing Hydra-Optuna integration?
  2. How should we handle different model types and specialized parameters?
  3. Should optimization be done at the full pipeline level or individual components?
  4. How to ensure reproductibility of optimization runs?

**Acceptance Criteria**

  - Documentation with examples for common use cases
  - Support for basic parameter types (int, float, categorical)
  - Integration with existing metrics for objective functions
  - MLflow logging of optimization results
  - Ability to export best configuration for further use

**Priority**

  Medium: Important for model optimization, but not blocking current development.
